tests/src/com.baomidou/dynamic-datasource-spring-boot-starter/3.6.1/build.gradle

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+
+plugins {
+    id "org.graalvm.internal.tck"
+    id 'org.springframework.boot' version '3.0.0'
+    id 'io.spring.dependency-management' version '1.1.0'
+}
+
+String libraryVersion = tck.testedLibraryVersion.get()
+
+dependencies {
+    testImplementation "com.baomidou:dynamic-datasource-spring-boot-starter:$libraryVersion"
+    testImplementation 'org.assertj:assertj-core:3.22.0'
+    testImplementation 'org.springframework.boot:spring-boot-starter-web'
+    testRuntimeOnly 'com.h2database:h2:2.1.214'
+    testImplementation 'org.springframework.boot:spring-boot-starter-test'
+}
+
+graalvmNative {
+    agent {
+        defaultMode = "conditional"
+        modes {
+            conditional {
+                userCodeFilterPath = "user-code-filter.json"
+            }
+        }
+        metadataCopy {
+            mergeWithExisting = true
+            inputTaskNames.add("test")
+            outputDirectories.add("src/test/resources/META-INF/native-image/com.baomidou/dynamic-datasource-spring-boot-starter")
+        }
+    }
+}
```

Comment:
@sdeleuze 
- Hi, I noticed that when Spring Boot's Gradle Plugin is introduced, the following error will occur when executing `./gradlew -Pagent clean test`.

```shell
01:21:29.638 [main] INFO org.springframework.test.context.aot.TestClassScanner - Scanning for Spring test classes in all packages in classpath roots [/home/linghengqian/TwinklingLiftWorks/git/public/graalvm-trace-metadata-smoketest/dynamic-datasource-spring-boot-starter/build/classes/java/test, /home/linghengqian/TwinklingLiftWorks/git/public/graalvm-trace-metadata-smoketest/dynamic-datasource-spring-boot-starter/build/resources/test]
Exception in thread "main" java.lang.IllegalStateException: Cannot perform AOT processing during AOT run-time execution
        at org.springframework.util.Assert.state(Assert.java:76)
        at org.springframework.test.context.aot.TestContextAotGenerator.processAheadOfTime(TestContextAotGenerator.java:126)
        at org.springframework.test.context.aot.TestAotProcessor.performAotProcessing(TestAotProcessor.java:91)
        at org.springframework.test.context.aot.TestAotProcessor.doProcess(TestAotProcessor.java:72)
        at org.springframework.test.context.aot.TestAotProcessor.doProcess(TestAotProcessor.java:39)
        at org.springframework.context.aot.AbstractAotProcessor.process(AbstractAotProcessor.java:82)
        at org.springframework.boot.test.context.SpringBootTestAotProcessor.main(SpringBootTestAotProcessor.java:57)
```
- If I just execute `./gradlew clean test`, everything works fine, but cannot use `user-code-filter.json` to generate GraalVM reachability metadata in Conditional Agent Mode. 
- Spring AOT overrides my configuration for Native Build Tools. I wonder if this is the expected behavior of https://github.com/spring-projects/spring-framework/issues/28163 ? Or am I overlooking something?

Reply:
@linghengqian Looks like Native Build Tools sets `org.graalvm.nativeimage.imagecode` system property when the agent mode is set (expected) but that impact the code running in the Gradle plugin JVM potentially because the JVM is not forked.

Any thoughts @bclozel @sbrannen?

Reply:
Running in agent mode with the NBT will set `org.graalvm.nativeimage.imagecode=agent`.

And `AotDetector.useGeneratedArtifacts()` returns `true` if that's the case.

So that would explain the cause of the `IllegalStateException` in `TestContextAotGenerator`.

That may be problematic in and of itself.

----

In any case, why is `TestContextAotGenerator` used when executing `./gradlew -Pagent clean test`?

Is that command supposed to trigger AOT processing of the test suite?

Reply:
@sbrannen 

- This involves an idea that I want to use GraalVM Tracing Agent to collect GraalVM Reachability Metadata in Conditional form for the corresponding unit tests for some third-party Spring Boot Starters that must use Spring Boot's unique classes.  

- I've been involved with too many deadlines lately, and haven't had time to investigate this PR.  

- Do you want me to open a new issue with a minimal reproducible example on the Spring Boot side?

Reply:
> Do you want me to open a new issue with a minimal reproducible example on the Spring Boot side?

@linghengqian, yes that would be great!

Though it may be better to file the issue against the Spring Framework, since changes will likely need to be made in Framework -- even if the example you provide uses Boot and its plugins.

Reply:
> > Do you want me to open a new issue with a minimal reproducible example on the Spring Boot side?
> 
> @linghengqian, yes that would be great!
> 
> Though it may be better to file the issue against the Spring Framework, since changes will likely need to be made in Framework -- even if the example you provide uses Boot and its plugins.

- @sbrannen  I'm sorry I've been busy and forgot about this. I'm open at https://github.com/spring-projects/spring-boot/issues/34841. If you think it is necessary, you can move the issue to the Spring Framework side, because I think this is only related to Spring's Gradle Plugin.